tests: drivers: uart: config api

jenmwms opened the pull request:
The UART API now supports configure and configure_get functions,
however no tests were written for those functions. This
commit adds the framework and implementing tests for configure
and configure_get for UART API.

Fixes (#12872)

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

jenmwms committed:
tests: drivers: uart: config api

The UART API now supports configure and configure_get functions,
however no tests were written for those functions. This
commit adds the framework and implementing tests for configure
and configure_get for UART API.

Fixes (#12872)

Signed-off-by: Jennifer Williams <jennifer.m.williams@intel.com>

stephanosio commented on tests/drivers/uart/uart_basic_api/src/test_uart_config.c:
```diff
@@ -0,0 +1,91 @@
+	while (1) {
+		while (uart_configure(uart_dev, &uart_cfg) < 0) {
+			/* returns 0 if successful, negative error code otherwise */
+		}
```
If `uart_configure` keeps returning an error code (i.e. negative value), the test will hang. Is this intentional?

Should "assert" here instead?

jenmwms replied:
Thanks for asking.  I wondered this too, I based this on the existing tests as reference in `test_uart_poll.c.` The while loop(s) are not needed here like in the polling case. I can remove it and make this a 1x call to `uart_configure( )`, and check if error or 0/success is returned.

The assert is done in `test_uart_configure( )` which calls this using `zassert_true( )`.

stephanosio commented on tests/drivers/uart/uart_basic_api/src/test_uart_config.c:
```diff
@@ -0,0 +1,91 @@
+	while (1) {
+		while (uart_config_get(uart_dev, &uart_cfg) < 0) {
+		/* returns 0 if successful, negative error code otherwise */
+		}
```
same problem here

jenmwms replied:
see previous comment

stephanosio commented on tests/drivers/uart/uart_basic_api/src/test_uart_config.c:
```diff
@@ -0,0 +1,91 @@
+		/* returns 0 if successful, negative error code otherwise */
+		}
+		break;
+	}
```
What is the purpose of this `while` loop?

SebastianBoe replied:
The same applies above as well.

jenmwms replied:
see previous comment.

jenmwms commented:
I removed the extraneous while loop(s) and simplified this to a 1x call to uart_configure( ) and uart_config_get(), and check if error or 0/success is returned.

jenmwms commented:
1 test failed in Shippable - 

Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/obj_tracing/src/main.c:113: object_monitor: (((thread_counter == (TOTAL_THREADS + initial_count)) && (obj_counter == TOTAL_OBJECTS)) is false)
test failed

stephanosio commented:
> 1 test failed in Shippable -
> 
> Assertion failed at WEST_TOPDIR/zephyr/tests/kernel/obj_tracing/src/main.c:113: object_monitor: (((thread_counter == (TOTAL_THREADS + initial_count)) && (obj_counter == TOTAL_OBJECTS)) is false)
> test failed

We currently have intermittent failure issues for some tests in the CI. I have re-triggered Shippable, so it should pass eventually.

stephanosio commented on tests/drivers/uart/uart_basic_api/src/test_uart_config.c:
```diff
@@ -0,0 +1,93 @@
+		return TC_PASS;
+	} else {
+		return TC_FAIL;
+	}
```
Nit: could do `return (ret == 0) ? TC_PASS : TC_FAIL;`

jenmwms replied:
I updated this. Much cleaner, thanks!

stephanosio reviewed (approved):
Looks ok to me

chen-png commented on tests/drivers/uart/uart_basic_api/src/test_uart_config.c:
```diff
@@ -0,0 +1,93 @@
+		.data_bits = UART_CFG_DATA_BITS_8,
+		.flow_ctrl = UART_CFG_FLOW_CTRL_NONE
+	};
+
```
do we still need to initialize this strcut uart_cfg?

jenmwms replied:
I updated this. Much cleaner, thanks!

xuhao8210 commented:
@jenmwms -- I have a small comment for the test cases, do we need to add one checking point to check whether output value is same with the input value for the uart config ?

If yes, we may merge the 2 test cases into one case , then add the checking point ,or create a new test case for that. What are your comments ? Thanks.

jenmwms commented:
@xuhao8210 Thanks for suggesting. That is a great improvement, I was remaining consistent with related test structure of test_uart_poll.c, which would need the same enhancement? Currently, these test the function call and return. I will update this test to be more complete.

jenmwms commented:
I added the comparison @xuhao8210, please take a look and see if you agree. 
Thanks!

xuhao8210 commented:
@jenmwms  ---- It seems that the 2nd case: ”test_config_get()” will fail if we don’t run the 1st case :” test_uart_configure()” because the 2nd case have not set the value for the variable :”uart_cfg_check”. 
So I suggest to make each case's independence as possible , and I have few draft ideas, could you please help review them below?  Thanks in advance.  

1. For the 1st case , we may only force on testing for setting configuration, so we may use the 1st case from your last version: 6c1156d to name test case: “test_uart_config_set()”.
2.  For the 2nd case, we may use the 1st case from your new version:  721ce3a as 2nd test case, then name test case:”test_uart_config_get()”, and I have few comments below for the case’s last part.
Could you please help also review the code below ? 

================last part code ==================
/* Verify configure() - set device configuration using data in cfg */
int ret = uart_configure(uart_dev, &uart_cfg);
zassert_ture(ret == 0, "set config error");
 /* Confirm the values provided are the values set*/
ret = uart_config_get(uart_dev, &uart_cfg_check);
 zassert_ture(ret == 0, "get config error");
if (memcmp(&uart_cfg, &uart_cfg_check, sizeof(uart_cfg)) != 0) {
                                         return TC_FAIL; 
                } else {
                     return TC_PASS;
                }
 ==============End last part code ==================

jhedberg commented:
@jenmwms @xuhao8210 I was a bit too hasty in merging this post release (it had required approvals and passed CI), so I didn't notice that there was still this conversation going on. Anyway, since it's merged you'll need to fix whatever remains through a follow-up PR.

xuhao8210 commented:
@jhedberg  ----OK. got it, and we will continue to follow up to discuss about this PR to get better solution.  Thanks.

jenmwms commented:
@jhedberg @xuhao8210 Thank you both! This PR is ok then.I opened an issue/enhancement #23459 to host the continued conversation and we can resolve it with a followup PR.
